keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -77,15 +77,15 @@ checklist items _must_ be updated for the enhancement to be released.
 Items marked with (R) are required *prior to targeting to a milestone /
 release*.
 
--   [ ] (R) Enhancement issue in release milestone, which links to KEP dir in
+-   [X] (R) Enhancement issue in release milestone, which links to KEP dir in
     [kubernetes/enhancements](not the initial KEP PR)
--   [ ] (R) KEP approvers have approved the KEP status as `implementable`
--   [ ] (R) Design details are appropriately documented
--   [ ] (R) Test plan is in place, giving consideration to SIG Architecture and
+-   [X] (R) KEP approvers have approved the KEP status as `implementable`
+-   [X] (R) Design details are appropriately documented
+-   [X] (R) Test plan is in place, giving consideration to SIG Architecture and
     SIG Testing input (including test refactors)
--   [ ] (R) Graduation criteria is in place
--   [ ] (R) Production readiness review completed
--   [ ] (R) Production readiness review approved
+-   [X] (R) Graduation criteria is in place
+-   [X] (R) Production readiness review completed
+-   [X] (R) Production readiness review approved
```

Comment:
Sorry - I can't comment on unchanged lines, so commenting here:

L226: Will this be a built-in API or a CRD?

L241: I suggest using NodeSelector instead of LabelSelector:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2643

L299: I think that we need to mention how we handle cases where there aren't any free IPs anymore.
So basically the algorithm should be sth like:
When trying to allocate a pod-cidr for a node:
(1) discard all ClusterCIDRConfigs where there aren't any free IPs or those that doesn't match the selector.
(2) From the remaining prefer ....
<--- update ---> I see that being discussed below - would be more intuitive to merge it to sth like I mentioned above.

L364: It might be a mental shortcut, but we should be clear - periodic polling doesn't scale well - the controller should be watching (and potentially polling from the local cache to check, though even that isn't necessary needed with our informer machinery).


L452: I think I'm not fully following - are you saying that: whether we apply IPv4, IPv6 or both is decided purely on their availability in the ClusterCIDRConfig object?
[If not can you clarify? If so, can we also make it more explicit?]

L500: if the IPs from the previously existing `created-from-flags-\<hash\>` are in-use, then finalizer will block its deletion. Does that block initialization of the new ipam controller? That sounds like a problem to me...

L515: This shouldn't be different from a regular operation - on failed operation we should simply be re-queuing (with backoff) the node to retry [we should need any additional logic.] (This pattern is used in couple other controllers).

L732: I would say that we should have a feature gate too. So basically:
- a feature-gate that is deciding if the new controller can even be started
- and on top of that a flag that decides which controller to use

L758: Not really restart nodes - but rather recreate them.

L773: nit: "No - the (...) tests will be added before graduating the feature to Alpha." 
[The tests doesn't yet exist, right?]

Reply:
I've made changes to reflect the comments above. I hope I clarified everything.

For L500: The bit that's probably getting lost is that the controller chooses a random hash each time it starts up. So if it needs to create a new object and delete an old one, it need not block. However, users will still have to recreate their Nodes -- there's no way around that.

Reply:
Thank you - that's perfect!